autoextract/aio/client.py

```diff
@@ -61,14 +155,19 @@ async def request_raw(query: Query,
 
     When handle_retries is True, this function can raise
 
-    1) autoextract.errors.ApiError, if there is an error returned by the API
-       which is not a throttling response (e.g. it can be raised for incorrect
-       request).
-    2) tenacity.RetryError, if a network-related error persists for
-       a long time, over the allowed time period.
+    1) autoextract.errors.RequestError,
+       if there is a Request-level error returned by the API
+       which is not a throttling response
+       (e.g. it can be raised for incorrect request).
+    2) tenacity.RetryError,
```

Comment:
Is that right? tenacity.RetryErrors are raised by `request_raw`? I think is not the case, but not sure. I think they shouldn't. Also, I think `RequestError` should also be raised in the case that `handle_retries` is False. Not sure if this the case...

Reply:
> Is that right? tenacity.RetryErrors are raised by request_raw? I think is not the case, but not sure.

Yes, that's true. See here:

```python
    try:
        # Try to make a batch request
        result = await request()
    except QueryRetryError:
        # If Tenacity fails to retry a QueryError because the max number of
        # retries or a timeout was reached, get latest results combining
        # error and successes and consider it as the final result.
        result = request_processor.get_latest_results()
    except Exception:
        agg_stats.n_fatal_errors += 1
        raise
```

The `tenacity.RetryError` exceptions are captured by the `except Exception` just to increment fatal error stats and then they are raised again. This was also the behavior before my pull request.

> I think they shouldn't. 

This is something we need to discuss, but it's probably out of scope. 
Any thoughts, @kmike and @ejulio?

> Also, I think RequestError should also be raised in the case that handle_retries is False. Not sure if this the case...

It is raised. What will happen is that since retries are disabled, Tenacity won't capture it and it will be re-raised per the `except Exception` block as described above.

```python
                if resp.status >= 400:
                    # ...
                    raise RequestError(
                        request_info=resp.request_info,
                        history=resp.history,
                        status=resp.status,
                        message=resp.reason,
                        headers=resp.headers,
                        response_content=content
                    )
```

Reply:
> The tenacity.RetryError exceptions are captured by the except Exception just to increment fatal error stats and then they are raised again. This was also the behavior before my pull request.

Ok, I think we have to discuss that, and I think it should be in the scope of this PR. My point is that we should never raise `tenacity.RetryError`. Instead, we should raise the underlying `RequestError` for request errors, or ` request_processor.get_latest_results()` otherwise. We could include the number of retrials in `RequestError` in case we want to make this information accessible by the client.

Reply:
To be done separately. Created this issue: https://github.com/scrapinghub/scrapinghub-autoextract/issues/11